### ISSUE TITLE
Separate publish / tag into two separate entrypoints

### ISSUE THREAD

**Issue by dfee** [atlassian/changesets#478]
## Affected Packages

pkg:@changesets/cli

## Problem

If `changeset publish` fails halfway through publishing, then no tags are created and pushed. If then, `changeset publish` succeeds, then tags are only created for packages who were published on that run. The problem, then, is then that we won't tag all releases, only the subset that were successfully published on the second run.

## Proposed solution

Split `tag` out of `publish`. `changeset version` -> `changeset publish` -> `changeset version`.

**Comment by emmenko** [comment]
I have a similar issue/use case.

One of our repositories has around ~90 packages. We noticed that when all (or most of the) packages need to be published, there is a higher chance of the publishing to fail. To be more specific, packages are correctly published to NPM but the changesets command somehow fails, which then prevents tags to be created (as reported in this issue).

For example you can see in the screenshot that the version is published but the command reports an error.

<img width="1531" alt="image" src="https://user-images.githubusercontent.com/1110551/98665296-c506ab00-234b-11eb-8ab3-2e8c5d9ddbff.png">

Do you have an idea what might cause this?

However, in general it would be good to have a more resilient "error strategy". For example, retrying the publishing but do not throw in case the package is indeed published, etc.

Thanks

**Comment by Andarist** [comment]
> Do you have an idea what might cause this?

We might want to limit concurrent requests to npm, even more, when publishing - I believe this has been discussed here: https://github.com/atlassian/changesets/issues/427#issuecomment-697206015

> However, in general it would be good to have a more resilient "error strategy". For example, retrying the publishing but do not throw in case the package is indeed published, etc.

That definitely would be an improvement. Additionally, to help the original issue reported here we could move creating tags at the end of the whole publish step. This should ensure that all packages are published first and only then we could attempt to create tags.

**Comment by dfee** [comment]
Tags are non-destructive, right? I.e. they're not going to be overwritten on the remote. Therefore, shouldn't tagging follow the same strategy as publishing? I.e. if it's not published, attempt to publish...

**Comment by Andarist** [comment]
Right, this probably could be done as well - it would probably require making an additional request to npm to check if a particular tag is published (unless we already get that info in the requests we make). Could you check what it would take to implement this?

**Comment by emmenko** [comment]
Can you point out which places in the code we should be looking into?
I can also try to help out if you need some help.

**Comment by dfee** [comment]
I'm generally thinking two paths:

### npm-registry path
1. enumerate current versions
2. identify which versions (are not) present on the NPM registry
3. push the deltas

### GitHub path
1. enumerate current versions
2. identify which tags (are not) present on GitHub
3. tag w/ a (provided) commit-id and push (or default to `HEAD`).

Thoughts?

**Comment by Andarist** [comment]
We hope that this change https://github.com/atlassian/changesets/pull/522 should result in fewer faulty publishes

**Comment by zhaoyao91** [comment]
Same requirement, different case. We want to use `changeset` to manage some small apps, not packages hosted in npm registry. The value is changeset workflow and management of changelog and tags, without `npm publish`. But currently, the only way to generate tag is to publish, which is not suitable.

**Comment by Andarist** [comment]
@zhaoyao91 I would accept a PR for this - this should be implemented as separate command

**Comment by mitchellhamilton** [comment]
I think it would probably be better if this was implemented in publish so that we always tag packages regardless of whether or not they're private and separately to the npm publish so if it's already published to npm, the tag will still be created.

If some one wants to do that, it would probably look like fixing up https://github.com/atlassian/changesets/pull/420 to do the tagging concurrently and do all the tagging in the same place, including for non-private packages.

**Comment by javier-garcia-meteologica** [comment]
`@changesets/cli` 2.18.0 ships with a separate `changeset tag` command, #612 #634

I think this issue is already solved

**Comment by bluwy** [comment]
Closed as `changeset tag` is supported. Pass `changeset publish --no-git-tag` to disable for the publish command.
